internal/incus/lxc_template_files.go

```diff
@@ -0,0 +1,53 @@
+package incus
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/lxc/incus/v6/shared/api"
+
+	"github.com/lxc/cluster-api-provider-incus/internal/static"
+)
+
+func (c *Client) ensureInstanceTemplateFiles(instanceName string) error {
+	metadata, _, err := c.Client.GetInstanceMetadata(instanceName)
+	if err != nil {
+		return fmt.Errorf("failed to GetInstanceMetadata: %w", err)
+	}
+
+	var mustUpdateMetadata bool
+	for _, file := range []struct {
+		templateName string
+		content      string
+		filePath     string
+		mode         string
+	}{
+		{templateName: "capn-install-kubeadm.tpl", filePath: "/opt/cluster-api/install-kubeadm.sh", content: static.InstallKubeadmScript(), mode: "0755"},
+		{templateName: "capn-kube-flannel.tpl", filePath: "/opt/cluster-api/kube-flannel.yaml", content: static.KubeFlannelTemplate(), mode: "0644"},
+		{templateName: "capn-kube-proxy-config.tpl", filePath: "/opt/cluster-api/kube-proxy-config-lxc.yaml", content: static.KubeProxyConfigTemplate(), mode: "0644"},
+		{templateName: "capn-kube-vip.tpl", filePath: "/opt/cluster-api/kube-vip-pod.yaml", content: static.KubeVIPTemplate(), mode: "0644"},
+	} {
+		if _, ok := metadata.Templates[file.filePath]; !ok {
+			if err := c.Client.CreateInstanceTemplateFile(instanceName, file.templateName, bytes.NewReader([]byte(file.content))); err != nil {
+				// TODO: do not fail if already exists
+				return fmt.Errorf("failed to CreateInstanceTemplateFile(%q): %w", file.templateName, err)
+			}
+
+			metadata.Templates[file.filePath] = &api.ImageMetadataTemplate{
+				When:       []string{"create", "copy", "start"},
+				CreateOnly: true,
```

Comment:
So that's a bit confusing, by setting `CreateOnly`, the template will only be run if the file doesn't already exist in the instance, which makes me wonder why have `When` set to `start` then given that it will only realistically only be run just once in which case having just `create` be the trigger should be sufficient.

But then if the template is meant to be re-applied on copy and or later during the instance's lifetime, the `CreateOnly` should probably be false at that point.

Reply:
Thanks for review, apparently `when: [create]` is enough here, since the templates are added before the instance starts for the first time.